src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -166,6 +168,30 @@ class SetupPySources:
     package_data: Tuple[PackageDatum, ...]
 
 
+@dataclass(frozen=True)
+class SetupPyKwargsRequest:
+    exported_target: ExportedTarget
+    requirements: ExportedTargetRequirements
+    sources: SetupPySources
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class SetupPyKwargs(FrozenDict):
+    """The keyword arguments to put in the setup.py file."""
+
+    json_str: str
+
+    def __init__(self, kwargs: Mapping[str, Any]) -> None:
+        super().__init__()
+        # Note that we convert to a `str` so that this type can be hashable.
```

Comment:
Probably want to sort keys, so the string is consistent? But also, why wouldn't it be hashable if we created a FrozenDict?

Reply:
I originally tried with `FrozenDict`, but it requires that every single part of the dictionary is immutable, e.g. that you don't have any `list` or `dict` as values. Turns out, we have a couple of those. I tried to convert them, but it ended up getting really clunky and it could break with dynamic values specified in a `BUILD` file.